File: src/test/java/io/github/mdsimmo/bomberman/game/ExplosionTest.java
package io.github.mdsimmo.bomberman.game;

import org.junit.Test;

import java.util.Map;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;

import static org.junit.Assert.assertEquals;

public class ExplosionTest {

    @Test
    public void lootSelectWithSingleItemAlwaysGetThatItem() {
        for (int i = 0; i < 100; i++) {
            Map<String, Number> loot = Map.of("diamond", 1);
            var got = Explosion.lootSelect(loot);

            assertEquals(Set.of("diamond"), got);
        }
    }

    @Test
    public void lootSelectWithNoLootGetsNothing() {
        for (int i = 0; i < 100; i++) {
            Map<String, Number> loot = Map.of();
            var got = Explosion.lootSelect(loot);

            assertEquals(Set.of(), got);
        }
    }

    @Test
    public void lootSelectIsWeightedUnbiased() {
        AtomicInteger diamonds = new AtomicInteger();
        AtomicInteger dirt = new AtomicInteger();
        var loot = Map.of(diamonds, 1, dirt, 3);

        for (int i = 0; i < 1000; i++) {
            var got = Explosion.lootSelect(loot);
            got.forEach(AtomicInteger::incrementAndGet);
        }

        assertEquals(diamonds.get(), 250, 30);
        assertEquals(dirt.get(), 750, 30);
    }
}

File: src/test/java/io/github/mdsimmo/bomberman/game/ExplosionTest.kt
package io.github.mdsimmo.bomberman.game

import io.github.mdsimmo.bomberman.game.Explosion.Companion.lootSelect
import org.junit.Assert
import org.junit.Test
import java.util.concurrent.atomic.AtomicInteger
import java.util.function.Consumer

class ExplosionTest {
    @Test
    fun lootSelectWithSingleItemAlwaysGetThatItem() {
        for (i in 0..99) {
            val loot = mapOf(Pair("diamond", 1))
            val got = lootSelect(loot)
            Assert.assertEquals(setOf("diamond"), got)
        }
    }

    @Test
    fun lootSelectWithNoLootGetsNothing() {
        for (i in 0..99) {
            val loot = mapOf<String, Number>()
            val got = lootSelect(loot)
            Assert.assertEquals(setOf<String>(), got)
        }
    }

    @Test
    fun lootSelectIsWeightedUnbiased() {
        val diamonds = AtomicInteger()
        val dirt = AtomicInteger()
        val loot = mapOf(
                Pair(diamonds, 1),
                Pair(dirt, 3)
        )

        for (i in 0..999) {
            val got = lootSelect(loot)
            got.forEach(Consumer { obj: AtomicInteger -> obj.incrementAndGet() })
        }

        Assert.assertEquals(diamonds.get().toFloat(), 250f, 40f)
        Assert.assertEquals(dirt.get().toFloat(), 750f, 40f)
    }
}